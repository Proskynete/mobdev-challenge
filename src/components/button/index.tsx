interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: "contained" | "outlined";
}

const Button = ({ variant = "contained", ...rest }: ButtonProps) => {
  return (
    <button
      {...rest}
      className={`${
        variant === "contained"
          ? "bg-blue-500 hover:bg-blue-700 text-white border-blue-500 border"
          : "bg-transparent hover:bg-blue-500 text-blue-700 hover:text-white border-blue-500 border hover:border-transparent"
      } py-2 px-4 rounded disabled:pointer-events-none disabled:opacity-50`}
    >
      {rest.children}
    </button>
  );
};

export { Button };

//
