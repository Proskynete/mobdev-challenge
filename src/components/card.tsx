import { useQuery } from "react-query";
import { DogAPI } from "../services/dog_api";

interface CardProps {
  name: string;
}

const Card = ({ name }: CardProps) => {
  const { data } = useQuery({
    queryKey: ["GET_DOG_IMAGE_BY_NAME", name],
    queryFn: () => DogAPI.getImageByName(name),
  });

  return (
    <div className="flex flex-col justify-center items-center bg-gray-800 rounded-lg p-4 hover:bg-gray-700 transition duration-200 cursor-pointer">
      <img
        src={data?.message}
        alt={data?.message}
        className="w-32 h-32 rounded-full"
      />
      <p className="text-center font-bold text-white mt-3">{name}</p>
    </div>
  );
};

export { Card };
