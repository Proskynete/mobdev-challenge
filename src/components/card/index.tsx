import { useQuery } from "react-query";
import { DogAPI } from "../../services/api";
import { useNavigate } from "react-router-dom";

interface CardProps {
  name: string;
}

const Card = ({ name }: CardProps) => {
  const navigate = useNavigate();

  const { data, isLoading } = useQuery({
    queryKey: ["GET_DOG_IMAGE_BY_NAME", name],
    queryFn: () => DogAPI.getImageByName(name),
  });

  const handleClick = () => {
    navigate(`/${name}`);
  };

  return (
    <div
      className="flex flex-col justify-center items-center bg-gray-800 rounded-lg p-4 hover:bg-gray-700 transition duration-200 cursor-pointer"
      onClick={handleClick}
    >
      {isLoading ? (
        <div className="w-32 h-32 rounded-full bg-gray-700 animate-pulse" />
      ) : (
        <img
          src={data?.message}
          alt={data?.message}
          className={`w-32 h-32 rounded-full object-cover object-center border-4 border-gray-700`}
          width="128"
          height="128"
          loading="lazy"
        />
      )}
      <p className="text-center font-bold text-white mt-3">{name}</p>
    </div>
  );
};

export { Card };
