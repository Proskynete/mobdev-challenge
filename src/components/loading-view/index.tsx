import { PropsWithChildren } from "react";

interface LoadingViewProps {
  loading: boolean;
  noResults: boolean;
  refetch?: (() => void) | boolean;
}

export const LoadingView = ({
  loading,
  noResults,
  children,
  refetch,
}: PropsWithChildren<LoadingViewProps>) => {
  return (
    <>
      {loading ? (
        <div className="flex flex-row justify-center items-center">
          <p className="text-center text-white">Cargando... üê∂</p>
        </div>
      ) : !loading && noResults ? (
        <div className="flex flex-row justify-center items-center">
          <p className="text-center text-white">
            No se encontraron resultados.
          </p>
        </div>
      ) : (
        <div className="flex flex-col justify-center items-center">
          {children}

          {!!refetch && (
            <div className="flex flex-row justify-center items-center mt-10">
              <button
                className="bg-gray-800 text-white font-bold py-2 px-4 rounded hover:bg-gray-700 transition duration-200"
                onClick={() => {
                  if (typeof refetch === "function") refetch();
                }}
              >
                Volver a buscar
              </button>
            </div>
          )}
        </div>
      )}
    </>
  );
};
