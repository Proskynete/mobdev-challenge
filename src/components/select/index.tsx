import { ChangeEvent } from "react";

interface SelectProps {
  id?: string;
  defaultLabel?: string;
  options: [string, string[]][] | string[];
  value?: string;
  onChange: (e: ChangeEvent<HTMLSelectElement>) => void;
}

const Select = ({
  defaultLabel,
  id,
  options,
  value,
  onChange,
}: SelectProps) => {
  return (
    <select
      id={id}
      className="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
      onChange={onChange}
      value={value}
    >
      <option value="">{defaultLabel || "Selecciona una opci√≥n"}</option>

      {options?.map((option) => {
        if (typeof option === "string") {
          return (
            <option key={option} value={option}>
              {option}
            </option>
          );
        }

        return (
          <option key={option[0]} label={option[0]}>
            {option[0]}
          </option>
        );
      })}
    </select>
  );
};

export { Select };
