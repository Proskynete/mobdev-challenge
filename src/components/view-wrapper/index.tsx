import { PropsWithChildren } from "react";

interface ViewWrapperProps {
  loading: boolean;
  noResults: boolean;
  refetch?: (() => void) | boolean;
}

const ViewWrapper = ({
  loading,
  noResults,
  children,
  refetch,
}: PropsWithChildren<ViewWrapperProps>) => {
  return (
    <>
      {loading ? (
        <div className="flex flex-row justify-center items-center">
          <p className="text-center text-white">Cargando... üê∂</p>
        </div>
      ) : !loading && noResults ? (
        <div className="flex flex-row justify-center items-center">
          <p className="text-center text-white">
            No se encontraron resultados.
          </p>
        </div>
      ) : (
        <div className="flex flex-col justify-center items-center">
          {children}

          {!!refetch && (
            <div className="flex flex-row justify-center items-center mt-10">
              <button
                className="bg-gray-800 text-white font-bold py-2 px-4 rounded hover:bg-gray-700 transition duration-200"
                onClick={() => {
                  if (typeof refetch === "function") refetch();
                }}
              >
                Volver a buscar
              </button>
            </div>
          )}
        </div>
      )}
    </>
  );
};

export { ViewWrapper };
