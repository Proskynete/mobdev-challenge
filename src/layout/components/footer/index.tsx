const Footer = () => {
  return (
    <div className="flex flex-col justify-center items-center bg-gray-800 py-4">
      <div className="flex flex-col mb-2 text-sm text-gray-500 text-center">
        <pre className="text-xs text-gray-800 dark:text-gray-100">
          &gt; $ cd ~/eduardoalvarez.dev/2024
          <span className="w-1 h-4 inline-block bg-primary-800 ml-2 rounded-sm motion-safe:animate-ping motion-safe:duration-75"></span>
        </pre>
        <div className="text-sm text-gray-500 dark:text-gray-400">
          <a href="https://eduardoalvarez.dev">eduardoalvarez.dev</a>
        </div>
      </div>
    </div>
  );
};

export { Footer };
