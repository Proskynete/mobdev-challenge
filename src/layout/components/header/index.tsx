import { ChangeEvent, useEffect, useState } from "react";
import { useBreed } from "../../../hooks/useBreed";
import { useNavigate, useParams } from "react-router-dom";
import { Select } from "../../../components/select";
import { Button } from "../../../components/button";

interface Filter {
  breed?: string;
  subBreed?: string;
}

const initialFilter: Filter = {
  breed: undefined,
  subBreed: undefined,
};

const Header = () => {
  const navigate = useNavigate();
  const { breed, subBreed } = useParams();

  const { allBreeds } = useBreed();
  const [subBreeds, setSubBreeds] = useState<string[] | null>(null);
  const [filter, setFilter] = useState<Filter>(initialFilter);

  useEffect(() => {
    if (breed) {
      setFilter((prev) => ({ ...prev, breed }));
      const subBreeds = allBreeds?.find(([name]) => name === breed)?.[1];
      setSubBreeds(subBreeds || null);
    }

    if (subBreed) setFilter((prev) => ({ ...prev, subBreed }));
  }, [breed, subBreed, allBreeds]);

  const handleSelectBreed = (e: ChangeEvent<HTMLSelectElement>) => {
    const breedSelected = e.target.value;
    setFilter((prev) => ({ ...prev, breed: breedSelected }));

    const subBreeds = allBreeds?.find(([name]) => name === breedSelected)?.[1];
    setSubBreeds(subBreeds || null);

    if (subBreeds?.length === 1) {
      setFilter((prev) => ({ ...prev, subBreed: subBreeds[0] }));
    }
  };

  const handleSelectSubBreed = (e: ChangeEvent<HTMLSelectElement>) => {
    const subBreed = e.target.value;
    setFilter({ ...filter, subBreed: subBreed });
  };

  const handleSearch = () => {
    if (!filter.breed && !filter.subBreed) {
      navigate("/");
    }

    if (filter.breed && !filter.subBreed) {
      navigate(`/breed/${filter.breed}`);
    }
    if (filter.subBreed && filter.breed) {
      navigate(`/breed/${filter.breed}/sub-breed/${filter.subBreed}`);
    }
  };

  const handleClear = () => {
    navigate("/");
    setFilter(initialFilter);
    setSubBreeds(null);
  };

  return (
    <div className="flex flex-col justify-center items-center bg-gray-800 px-6 py-4">
      <h1 className="text-2xl font-bold text-white">mobdev challenge!</h1>
      <p className="text-center text-white">
        Encuentra im√°genes (en forma aleatoria) de perros por raza y sub-raza.
      </p>

      <div className="w-full max-w-xl mt-6 flex flex-col gap-4 md:flex-row md:justify-between md:items-center">
        <p className="text-white">FILTROS:</p>

        <Select
          id="breeds"
          defaultLabel="Selecciona una raza"
          options={allBreeds}
          value={filter.breed || ""}
          onChange={handleSelectBreed}
        />

        <Select
          id="subBreeds"
          defaultLabel="Selecciona una sub-raza"
          options={subBreeds || []}
          value={filter.subBreed || ""}
          onChange={handleSelectSubBreed}
          disabled={!filter.breed}
        />

        <Button onClick={handleSearch} disabled={!filter.breed}>
          Buscar
        </Button>

        <Button
          variant="outlined"
          onClick={handleClear}
          disabled={!filter.breed || !filter.subBreed}
        >
          Limpiar
        </Button>
      </div>
    </div>
  );
};

export { Header };
