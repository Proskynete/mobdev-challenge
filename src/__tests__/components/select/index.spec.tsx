import { fireEvent, render, screen } from "@testing-library/react";
import { Select } from "../../../components/select";

describe("components/select", () => {
  it("should render component with default props - create snapshot", () => {
    const component = render(<Select options={["Option 1", "Option 2"]} />);

    expect(component).toMatchSnapshot();
  });

  it("should render select with provided default label", () => {
    render(
      <Select
        defaultLabel="Selecciona una opción"
        options={["Option 1", "Option 2"]}
      />
    );

    expect(screen.getByText("Selecciona una opción")).toBeInTheDocument();
  });

  it("should render provided options", () => {
    render(<Select options={["Option 1", "Option 2"]} />);

    expect(screen.getByText("Option 1")).toBeInTheDocument();
    expect(screen.getByText("Option 2")).toBeInTheDocument();
  });

  it("should render provided options with labels", () => {
    render(
      <Select
        options={[
          ["Label 1", ["Option 1"]],
          ["Label 2", ["Option 2"]],
        ]}
      />
    );

    expect(screen.getByText("Label 1")).toBeInTheDocument();
    expect(screen.getByText("Label 2")).toBeInTheDocument();
  });

  it("should call onChange when an option is selected", () => {
    const handleChange = jest.fn();
    render(
      <Select options={["Option 1", "Option 2"]} onChange={handleChange} />
    );

    fireEvent.change(screen.getByRole("combobox"), {
      target: { value: "Option 1" },
    });
    expect(handleChange).toHaveBeenCalledTimes(1);
  });
});
