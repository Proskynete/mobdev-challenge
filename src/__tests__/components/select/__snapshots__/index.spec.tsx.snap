// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/select should render component with default props - create snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <select
        class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
      >
        <option
          value=""
        >
          Selecciona una opción
        </option>
        <option
          value="Option 1"
        >
          Option 1
        </option>
        <option
          value="Option 2"
        >
          Option 2
        </option>
      </select>
    </div>
  </body>,
  "container": <div>
    <select
      class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
    >
      <option
        value=""
      >
        Selecciona una opción
      </option>
      <option
        value="Option 1"
      >
        Option 1
      </option>
      <option
        value="Option 2"
      >
        Option 2
      </option>
    </select>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
